### PR TITLE
test(chart): Sanity tests Selenium Grid chart via Makefile commands

### LINKED ISSUE #1975: Helm Charts maintainer needed
### Feature and motivation

We have been unable to handle updates and proper testing for the helm chart. We are looking for someone who wants to maintain them long-term.

### Usage example

If we do not find someone by the end of November 2023, we will move the chart to the https://github.com/seleniumhq-community organization.

### PR DESCRIPTION
**Pull request by VietND96**
**Thanks for contributing to the Docker-Selenium project!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://selenium.dev/documentation/en/contributing/) guidelines, applied for this repository.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description
Sanity tests Selenium Grid chart via Makefile commands

### Motivation and Context
Take advantage from PR #2027, there are some enhancements on sanity tests for the Selenium Grid chart (link to #1975)
1. `Makefile` is updated new `make` commands for a few purposes. Someone is able to test their changes in their local before pushing and running workflow. Try to have the same environment in development and CI
```
# Setup Kubernetes environment
make chart_setup_env

# Setup Kubernetes cluster
make chart_cluster_setup

# Test Selenium Grid on Kubernetes
make chart_test

# Cleanup Kubernetes cluster
make chart_cluster_cleanup
```
2. Add `charts/selenium-grid/TESTING.md` to capture all related testing to chart

3. Update chart test workflow to use `make` commands.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://selenium.dev/documentation/en/contributing/) document.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by VietND96** [comment]
Hi @amardeep2006, can you please also have a review on this? Thanks!

File: .github/workflows/helm-chart-test.yml [hunk 1]
```diff
@@ -13,105 +13,46 @@ permissions:
...
 
 jobs:
-  lint-test:
-    name: "Lint Tests with ct"
-    runs-on: ubuntu-latest
+  build-and-test:
+    # Skip job based on the commit message, only works in push to branches for now
+    if: contains(toJson(github.event.commits), '[skip ci]') == false
+    name: Build & test Docker images with Helm charts
+    runs-on: ubuntu-20.04
```
Comment (diemol) [inline .github/workflows/helm-chart-test.yml]: Why `ubuntu-20.04`?

File: .github/workflows/helm-chart-test.yml [hunk 2]
```diff
@@ -13,105 +13,46 @@ permissions:
...
-          export RUN_IN_DOCKER_COMPOSE=true
-          ./tests/bootstrap.sh ${{matrix.browser}}
\ No newline at end of file
+      - uses: actions/checkout@v4
+      - name: Output Docker info
+        run: docker info
+      - name: Set up Python 3.8
+        uses: actions/setup-python@v4.7.1
+        with:
+          python-version: 3.8
```
Comment (diemol) [inline .github/workflows/helm-chart-test.yml]: I think we can use at least 3.11

**Review by diemol — CHANGES_REQUESTED** [review]
Some minor comments.

File: .github/workflows/helm-chart-test.yml (hunk 1 above)
Comment (VietND96) [reply .github/workflows/helm-chart-test.yml]: I refer to the workflow of docker image build & test, so pick up this node OS.
Do I need to do a bulk update for other workflow to use the same node `ubuntu-latest`?

File: .github/workflows/helm-chart-test.yml (hunk 2 above)
Comment (VietND96) [reply .github/workflows/helm-chart-test.yml]: Sure, will update this together with node OS above.

File: .github/workflows/helm-chart-test.yml (hunk 1 above)
Comment (VietND96) [reply .github/workflows/helm-chart-test.yml]: > Why `ubuntu-20.04`?

I have done the bulk updates for all workflow to use the same `ubuntu-latest`

File: .github/workflows/helm-chart-test.yml (hunk 2 above)
Comment (VietND96) [reply .github/workflows/helm-chart-test.yml]: > I think we can use at least 3.11

There are 3 workflow are using this action, those are updated to `3.11`

**Review by diemol — APPROVED** [review]
Thank you, @vietnd96!

**Comment by amardeep2006** [comment]
> Hi @amardeep2006, can you please also have a review on this? Thanks!

1. I was looking at the run logs and seems few tests are failing with error 504. Probably Kubernetes needs some time to make the Browser Nodes operational.
2. Also It seems that even after failing the tests, the GitHub action does not break. Shouldn't it have failed ?
3. Is this possible to make kubernetes/helm/kubectl installation optional ? There may be developer who have kubernetes already up an running on the machine (e.g. I use Rancher desktop mostly that comes with helm,kubectl, ingress and k3s). 
They can simply apply chart and run ct lint / Selenium test cases.

https://github.com/SeleniumHQ/docker-selenium/actions/runs/7017524980/job/19090970950

**Comment by VietND96** [comment]
Yes, I was also aware of this flaky and incorrect behavior. I am trying to make it stable

**Comment by VietND96** [comment]
> Is this possible to make kubernetes/helm/kubectl installation optional ? There may be developer who have kubernetes already up an running on the machine (e.g. I use Rancher desktop mostly that comes with helm,kubectl, ingress and k3s).
They can simply apply chart and run ct lint / Selenium test cases.

On this use case, how about if you jump to run `make chart_test`? Is it working?
Later I will break down the script with individual tool/component install...

**Comment by amardeep2006** [comment]
> Yes, I was also aware of this flaky and incorrect behavior. I am trying to make it stable

I was using K8sSmokeTest.py for this. Basically it makes sure cluster is up an running before I fired the tests.

**Comment by VietND96** [comment]
> I was using K8sSmokeTest.py for this. Basically it makes sure cluster is up an running before I fired the tests.

Yes, while integrating autoscaling & ingress tests, I also tried to merge your script into using SmokeTest.py, since it is same func to wait for Hub is up, just some timeout values are pulled out as arguments, K8s tests set timeout a little bit longer due to scaling interval and time for node replicas from 0 to 1.
Regarding the 504 error, I found looks like same as https://github.com/kubernetes/ingress-nginx/issues/3976. The default timeout is 60s, if a request in the session queue over 60s without node serve on time, the client web driver will throw this error.
In the override values of ingress, under `ingress.annotations` I added
```yaml
    nginx.ingress.kubernetes.io/proxy-connect-timeout: "360"
    nginx.ingress.kubernetes.io/proxy-read-timeout: "360"
    nginx.ingress.kubernetes.io/proxy-send-timeout: "360"
```
Let us monitor how it is stable

Regarding incorrect behavior, failed tests without job failed, it is due to in bash script handle trapping the error signal inproperly, step to cleanup between return 0 replaces the first error code from tests.
All those things will be fixed in the next PR

**Pull request merged**